### ISSUE TITLE
Ignore stationary objects after first notification (e.g. a parked car)

### ISSUE THREAD

**Issue by ifermon** [danecreekphotography/node-deepstackai-trigger#341]
First of all, this project is awesome. Thank you for putting it together. And I have never seen such clear, well documented code. I have not worked with angular but I had no problem following the flow.

I am experiencing issues where, for example, a car comes over and parks in my driveway. It sits there. The cool down timer works perfectly and I only get the initial alert, however, if a second car, or a person for that matter, also comes I don't get a second alert. Or alternatively, nothing happens, the cool down time is exceeded, and some other leaf motion triggers a still, the AI detects the same car, and I get another alert about the car. 

Where I want to get to is the the same object type in more or less the same location doesn't trigger repeated alerts AND doesn't stop new alerts for new objects (subject to cool down), or new instances of the same type of object, from triggering a new alert. Basically the cool down timer is still in effect but it just ignores anything it has already detected.  People moving around will extend the cool down timer because they are "new" but the parked car will not keep extending the timer.

I am thinking this might work logically:
- for each trigger, keep a list of last seen objects / locations (which you do already)
- if an object of the same type / location exists in new trigger, ignore it always (i.e. it doesn't trigger)
- if there is an object / location in the list and it is not found in new trigger, delete it from list
- if there is a new object / location combo AND cool down timer is exceeded then generate alert
- if cool down timer is not exceed for trigger then don't send alert (normal process)

I think this would end up with the car coming to the house and parking in the driveway and an alert being triggered. Subsequent people moving  around (e.g. getting out of the car) won't trigger because the cool down timer isn't exceeded. But when they are in the house and the car is sitting there the car alone won't ever generate a new alert until / unless it moves because it will just keep getting detected and is already in the list. However, new people or the people leaving the house, cars driving by or new cars will all generate new alerts (assuming cool down period is exceeded). 

You'd have to put a bit of fudge factor in the location (maybe 5%?) so that movement of a few pixels is ignored, and we might have issues when it detects a truck one time and a car the next from the same picture (I don't know why, but I have seen it happen).  

So basically it looks like we'd need a new method named something like objPreviouslyDetected(filename, prediction) called from the isTriggered method - and we would negate the return value. And we'd need to remove objects from the list after we've gone through all the predictions if they weren't there. 

I'd be happy to take a stab at it but I'm a bit worried about not living up to your coding style - like I said, it is super clean and easy to understand.

**Comment by danecreekphotography** [comment]
This has come up before and the recommended solution is to configure two triggers. The first detects only cars but masks out the area where you park the car. The second only detects people.

**Comment by danecreekphotography** [comment]
Thanks for the kind comments on the code. I put a lot of effort in to keep it tidy so I can keep track of what the hell I wrote when I look at it in the future 😂

**Comment by ifermon** [comment]
That works if it's your car. But if you have company and they park in your
driveway then it doesn't work.

On Wed, Jul 8, 2020 at 4:04 PM danecreekphotography
notifications-at-github.com |github-ifermon| <xnvxgdaa5vdukxt@sneakemail.com>
wrote:

**Comment by danecreekphotography** [comment]
Why wouldn't it work? The AI system doesn't car whose car it is. As long as your camera has a view of both the driveway and the access to the driveway (e.g. the turn into it from the street) it should be fine.

The car trigger will detect the car as it passes along the access to the driveway (e.g. along the street, turning into the driveway). The entire driveway gets masked out as a no-detection zone for cars, so any car parked anywhere in it won't cause repeated triggers to fire.

The second trigger for people doesn't have a mask and will fire any time a person is in frame, including on the driveway surface.

**Comment by danecreekphotography** [comment]
And a general philosophical comment on the feature request: any kind of logic that involves that many conditions and saving state between triggers is... challenging to implement cleanly and without bugs. Even just the simple cooldown timer was a giant pain to implement, and all that had to remember was a single "what was the last time of the last trigger" state.

In general I prefer to find solutions to problems like this that can leverage the existing features. Something along the lines of your proposal is just very difficult to do well.

**Comment by ifermon** [comment]
You are right in that case. It would work for the driveway to split and mask it like that. Not so much for my daughter sun bathing in the back yard since there isn't really a defined area that she goes to. I completely understand the concern about complexity - let me branch it off and take a stab- you can take a look when I'm done (if I finish and it works that is) and decide then if you like it or not. It should really be called something like object persistence.

**Comment by danecreekphotography** [comment]
That's the fun of open source :) Instead of a branch though you should make a fork of the repo and then do the branch in your fork. Definitely give it a go and see what you come up with, and let me know if you have questions about where to insert things in the code or have trouble getting up and running. 

Here is where the check for cooldown time happens:

https://github.com/danecreekphotography/node-deepstackai-trigger/blob/e3bda37bf462018816f56dba4c62af0b85781124/src/Trigger.ts#L111

You will likely want to insert right after that a call to your function that determines whether the persisted object exists. Then write your `checkForPersistedObject()` method somewhere else in the class.

I strongly recommend you use Visual Studio Code as your editor along with the Remote - Containers extension for development. You should be able to clone the repo, open it in Visual Studio Code, then open it as a Remote Container and be ready to go. F5 will compile and run complete with sample images.

**Comment by ifermon** [comment]
So super helpful hints. I've got VS up and running w/ the code in the
container. Now a super newbie question - what language is it written in? A
TS file should be typescript, but all the tutorials I've seen so far are
full of "let" keywords and I don't see a single let in your code.

On Wed, Jul 8, 2020 at 7:00 PM danecreekphotography
notifications-at-github.com |github-ifermon| <xnvxgdaa5vdukxt@sneakemail.com>
wrote:

**Comment by danecreekphotography** [comment]
It's 100% TypeScript. `const` is used nearly everywhere instead of `let` since 99% of the time the variables get assigned once and never changed again.

**Comment by ifermon** [comment]
So the code changes took me about 1 hour, including giving myself a primer on typescript. I've never coded in js so I was starting from scratch, but it's not too hard. And again, I've got to say how awesome your code is to read.  Other than supporting changes (e.g. a new attribute in the trigger.yaml file) and comments it was about 5 lines of code. 

Once I got the coding done it took me hours, many hours, to get it working in docker. The code wasn't the issue, I just couldn't figure out how to get it built and deployed on my machine. Finally got it goin this morning and it's been running all day with good results. Generally it ignores objects after finding them the first time until they disappear (or move significantly). 

Take a look if you are interested. I'm also not super happy with my method name, so if you have a suggestion that makes more sense let me know.  You set a high bar.

**Comment by danecreekphotography** [comment]
I took a peek. From a structure/concept standpoint it looks good. I like it and can see the value. In order to proceed you'll need to:

1. Put the changes in a branch. It looks like you made the branch but didn't switch to it, so all your changes merged into the `main` branch of your fork instead of the branch.
2. Only check in files that relate to the change. You checked in changes to the settings files and whatnot with your personal config that shouldn't be included.
3. Open a pull request. Then I can go in and add a bunch of comments on things like variable names, error, handling, and some places where you'll need to add some additional code to round out the setting.

**Comment by ifermon** [comment]
Dane-

I should be able to tackle it this weekend. I'll try my best not to make a
mess of it. :)

On Thu, Jul 9, 2020 at 1:53 PM danecreekphotography
notifications-at-github.com |github-ifermon| <xnvxgdaa5vdukxt@sneakemail.com>
wrote:

**Comment by ifermon** [comment]
I clearly didn't get to it this weekend. I haven't forgotten about it, just
trying to carve out some time. The changes are easy, it's the testing that
I need to figure out. I have not worked with docker before and I'm having a
tough time figuring out how to get my changes to run in a container. I
thought docker-compose up would get it running but it doesn't seem to
always pick up my changes unless I go and delete the existing docker
containers.

**Comment by danecreekphotography** [comment]
For testing what I do is run from Visual Studio with the dev containers (which makes it super easy to get things up and running), then manually drop images from my security camera into the mounted sampleConfiguration folder.

**Comment by coby** [comment]
Hi @ifermon and @danecreekphotography, I'm checking in on this to see if there has been any progress on this. Sorry if it's already integrated and I overlooked it.

@danecreekphotography Thank you for your hard work on this project. It's awesome.

**Comment by ghzgod** [comment]
The only way I have been able to resolve this, is to have nodered read MQTT messages.  I use an RBE node which only allows messages to pass if they are **different** from previous messages.  When there is motion on my driveway, my car gets highlighted even though it is not the motion originator.  So if the bounding box coordinates change (via the RBE node) then my camera triggers recording.  But I think we can come up with something more elegant eventually.

**Comment by danecreekphotography** [comment]
Closing old issues.
